File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int get_leading_digits(long card_number);
int get_card_length(long card_number);
int luhn(long card_number);
void print_card_type(int leading_digits, int card_length, int checksum);

int main(void)
{
    long card_number = get_long("Enter credit card number: ");
    int leading_digits = get_leading_digits(card_number);
    int card_length = get_card_length(card_number);
    int checksum = luhn(card_number);

    print_card_type(leading_digits, card_length, checksum);
}

int get_leading_digits(long card_number)
{
    while (card_number >= 100)
    {
        card_number /= 10;
    }

    return (int) card_number;
}

int get_card_length(long card_number)
{
    int length = 0;

    while (card_number > 0)
    {
        card_number /= 10;
        length++;
    }

    return length;
}

int luhn(long card_number)
{
    int sum = 0;
    int is_second_digit = 0;

    while (card_number > 0)
    {
        int digit = card_number % 10;

        if (is_second_digit)
        {
            digit *= 2;
            sum += digit / 10 + digit % 10;
        }
        else
        {
            sum += digit;
        }

        card_number /= 10;
        is_second_digit = !is_second_digit;
    }

    return sum % 10;
}

void print_card_type(int leading_digits, int card_length, int checksum)
{
    if (checksum != 0)
    {
        printf("INVALID\n");
        return;
    }

    if ((leading_digits == 34 || leading_digits == 37) && card_length == 15)
    {
        printf("AMEX\n");
    }
    else if ((leading_digits >= 51 && leading_digits <= 55) && card_length == 16)
    {
        printf("MASTERCARD\n");
    }
    else if ((leading_digits / 10 == 4) && (card_length == 13 || card_length == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return;
}

File: Week6/sentimental-credit/credit.py
import re
from cs50 import get_string


def main():
    card_number = get_string("Number: ")

    if not re.match(r"^\d+$", card_number):
        print("INVALID")
        return

    if is_valid(card_number):
        print_card_type(card_number)
    else:
        print("INVALID")


def is_valid(number):
    total_sum = 0
    number_str = str(number)
    length = len(number_str)

    for i in range(length - 2, -1, -2):
        digit = int(number_str[i]) * 2
        total_sum += digit // 10
        total_sum += digit % 10

    for i in range(length - 1, -1, -2):
        total_sum += int(number_str[i])

    return total_sum % 10 == 0


def print_card_type(number):
    if re.match(r"^3[47][0-9]{13}$", number):
        print("AMEX")
    elif re.match(r"^5[1-5][0-9]{14}$", number):
        print("MASTERCARD")
    elif re.match(r"^4[0-9]{12}([0-9]{3})?$", number):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
